mono/metadata/cominterop.c

```diff
@@ -643,11 +648,24 @@ void
 mono_mb_emit_cominterop_get_function_pointer (MonoMethodBuilder *mb, MonoMethod *method)
 {
 #ifndef DISABLE_JIT
+	int slot;
+	gboolean error;
 	// get function pointer from 1st arg, the COM interface pointer
 	mono_mb_emit_ldarg (mb, 0);
-	mono_mb_emit_icon (mb, cominterop_get_com_slot_for_method (method));
-	mono_mb_emit_icall (mb, cominterop_get_function_pointer);
-	/* Leaves the function pointer on top of the stack */
+	slot = cominterop_get_com_slot_for_method (method, &error);
+	if (!error) {
+		mono_mb_emit_icon (mb, slot);
+		mono_mb_emit_icall (mb, cominterop_get_function_pointer);
+		/* Leaves the function pointer on top of the stack */
+	}
+	else {
+		/*
+		 * This code should never be executed and all invalid COM objects should be sieved out during their creation.
+		 * However, it will be emitted as a safeguard for cases when a method tries to interact with an invalid COM object from an unreachable basic block.
+		 */
+		char *msg = g_strdup_printf ("Unable to locate the interface on which method %s is defined.", mono_method_get_full_name (method));
+		mono_mb_emit_exception (mb, "SystemException", msg);
```

Comment:
This looks like the wrong exception, how did you come to use it?

Reply:
You're right, `System.Runtime.InteropServices.InvalidComObjectException` should suit it better, because invalid COM objects are what the code is meant to handle.
There's one caveat though which makes me a bit uncertain. Namely, this branch is never really meant to be executed since all invalid COM objects be spotted during their creation, so in this regard I think it would make better sense to throw `System.ExecutionEngineException`.
I'm curious what you think about it.